https://issues.redhat.com/browse/ACM-7151 Add ZTP procedure (#5578)
open-cluster-management/rhacm-docs: clusters/cluster_lifecycle/create_cluster_on_prem.adoc

Continue with xref:../cluster_lifecycle/access_cluster.adoc#accessing-your-cluster[Accessing your cluster] for instructions for accessing your cluster. 

[#on-prem-boot-host-bmh]

> == Booting a host address by using bare metal hosts for on-premise environments

Reviewer: I would move this to the new host inventory section, maybe with a title "Automation for adding bare metal hosts to the host inventory".  Also, this describes how to do it with the OpenShift Container Platform CLI.  We can also do it with the console.  Would that be a separate section or would we add it here?

Reviewer: For the console, it's pretty simple.  From the host inventory, click on “Add hosts” and select "With BMC Form".  Here, fill in the information about the host: Name, hostname, BMC address, Boot NIC MAC address, and username and password.  Then click "Create".

> Complete the following steps to use an automated procedure for booting a host address by using bare metal hosts:

Reviewer: ```suggestion
You can automate booting the Discovery Image of your infrastructure environment by having the baremetal-operator communicate with the BMC of each bare metal host.  The procedure involves creating a BareMetalHost resource and associated BMC secret for each host.  The automation is enabled by a label on the BareMetalHost that references the infrastructure environment.

The automation performs the following actions:
- Boots each bare metal host with the Discovery Image represented by the infrastructure environment.
- Reboots each host with the latest Discovery Image in case the infrastructure environment or any associated network configurations is updated.
- Upon discovery, associates each Agent resource with its corresponding BareMetalHost resource.
- Updates Agent resource properties based on information from the BareMetalHost, such as hostname, role, and installation disk.
- Approves the Agent for use as a cluster node.

Complete the following steps to use an automated procedure for booting a host by using the baremetal-operator:
```

Reviewer: Link to baremetal-operator docs for more information

. Create a BMC secret by applying the following YAML content and replacing values where needed:
+
[source,YAML]
----
apiVersion: v1
kind: Secret
metadata:
  name: <bmc-secret-name>

>   namespace: <namespace>

Reviewer: This must be the same namespace as the InfraEnv.

type: Opaque
data:
  username: <username>
  password: <password>
----

. Create a bare metal host by applying the following YAML content and replacing values where needed:
+
[source,YAML]
----
apiVersion: metal3.io/v1alpha1
kind: BareMetalHost
metadata:
  name: <bmh-name>
  namespace: <namespace>
  annotations:

>     inspect.metal3.io: disabled

Reviewer 2: This no longer applies for the converged flow, right?

Reviewer: Correct, our controller will set this annotation as necessary

  labels:
    infraenvs.agent-install.openshift.io: <your-infraenv> <1>
spec:

>   online: true
>   automatedCleaningMode: disabled <2>
>   bootMACAddress: <your-mac-address>  <3>
>   bmc:
>     address: <machine-address> <4>
>     credentialsName: <bmc-secret-name> <5>

Reviewer: ```suggestion
  online: true
  automatedCleaningMode: disabled <2>
  bootMACAddress: <your-mac-address>  <3>
  bmc:
    address: <machine-address> <4>
    credentialsName: <bmc-secret-name> <5>
  rootDeviceHints:
    deviceName: /dev/sda <6>
```

----

> <1> The value must match the InfraEnv name and exist in the same namespace.

Reviewer: I would comment on the "namespace" property on line 121 that it should match the namespace of the InfraEnv.

<2> If you do not set a value, the `metadata` value is automatically used.

> <3> Use your mac address.

Reviewer 3: this part is important, it should match mac address of one of the interfaces on the host, otherwise the integration with the infrastructure operator will not work.

Reviewer 2: +1, "your MAC address" is somewhat vague

Reviewer: Is this any MAC address?  Or specifically the one booting from?

> <4> Use your machine address.

Reviewer 2: Same. This is the address of the BMC plus the driver information. Maybe refer to the IPI docs.

<5> Make sure that the `credentialsName` value matches the name of the BMC secret you created.

> 

Reviewer: ```suggestion
<6> Optionally select the installation disk.  The available root device hints are described [here](https://docs.openshift.com/container-platform/4.13/post_installation_configuration/bare-metal-configuration.html).  Once the host is booted with the Discovery Image and the corresponding Agent resource is created, the installation disk will be set according to this hint.

```

> After turning on the host, the image starts downloading. This might take a few mintues. When the host is discovered, an agent custom resource is created automatically.

Reviewer 4: Note that per https://github.com/openshift/assisted-service/pull/2032, few images will add a 1 minute cool down period. This is to ensure it uses the very latest image in a case of multiple updates (which regenerate the image).

Reviewer: I don't think this is something we need to add to the user docs, unless I missed something?

> After turning on the host, the image starts downloading. This might take a few mintues. When the host is discovered, an agent custom resource is created automatically.

Reviewer: We should also mention that the Agent resource is also automatically approved in case the host was booted via BMH.

> See the following table to learn about annotations that are available for the agent custom resource:

Reviewer: These annotations are for the BareMetalHost, not the Agent.  The user puts them on the BMH, and then CIM copies the values to the spec of the corresponding Agent resource.

This is necessary for automated flows (e.g., ZTP) where the user creates the InfraEnv, BMHs, and ClusterDeployment+AgentClusterInstall.  They can't set Agent properties because the Agent CRs are created later in the process.  So they set the properties on the BMH and then we copy them to the Agents later.

|===

> | Annotation | Optional or required | Description

Reviewer: Maybe instead of a description, write that it sets the Agent property with the same name accordingly?

| `bmac.agent-install.openshift.io/role`

> | Required

Reviewer: @filanov Why is role required?  I don't think any are required
